File: src/main/java/frc/robot/util/PoseUtil.java
package frc.robot.util;

import edu.wpi.first.math.geometry.Pose2d;
import frc.robot.util.FieldConstants.Side;
import java.util.List;
import java.util.Map;

public class PoseUtil {
  public static Map<Pose2d, List<Pose2d>> branchPoses;

  public static int getClosestDesiredBranchID(Pose2d robotPose, Side side) {
    // Pose2d closestFace = robotPose.nearest(FieldConstants.getReefFaces());
    // int faceID = FieldConstants.getReefFaces().indexOf(closestFace);
    // return FieldConstants.getBranchID(faceID, side);

    List<Pose2d> allL4Poses =
        FieldConstants.getAllL4Poses().stream().map(pose -> pose.toPose2d()).toList();
    int sideIndex = side == Side.RIGHT ? 0 : 1;
    List<Pose2d> allL4PosesOnSide =
        java.util.stream.IntStream.range(0, allL4Poses.size())
            .filter(index -> index % 2 == sideIndex)
            .mapToObj(allL4Poses::get)
            .toList();
    System.out.println(allL4Poses.size());
    System.out.println(allL4PosesOnSide.size());
    return allL4Poses.indexOf(robotPose.nearest(allL4PosesOnSide));
  }

  public static int getBranchIDFromPose(Pose2d branchPose) {
    List<Pose2d> allL4Poses =
        FieldConstants.getAllL4Poses().stream().map(pose -> pose.toPose2d()).toList();
    Pose2d closestL4Pose = branchPose.nearest(allL4Poses);
    return allL4Poses.indexOf(closestL4Pose);
  }
}

File: src/main/java/frc/robot/util/PoseUtil.kt
package frc.robot.util

import edu.wpi.first.math.geometry.Pose2d
import frc.robot.util.FieldConstants.Side

object PoseUtil {
    // Cache the L4 poses to avoid recalculating them
    private val allL4Poses by lazy { FieldConstants.getAllL4Poses().map { it.toPose2d() } }
    
    // Cache side-specific poses
    private val rightSidePoses by lazy { allL4Poses.filterIndexed { index, _ -> index % 2 == 0 } }
    private val leftSidePoses by lazy { allL4Poses.filterIndexed { index, _ -> index % 2 == 1 } }

    fun getClosestDesiredBranchID(robotPose: Pose2d, side: Side): Int {
        val posesOnSide = if (side == Side.RIGHT) rightSidePoses else leftSidePoses
        val closestPose = robotPose.nearest(posesOnSide)
        return allL4Poses.indexOf(closestPose)
    }

    fun getBranchIDFromPose(branchPose: Pose2d): Int {
        val closestL4Pose = branchPose.nearest(allL4Poses)
        return allL4Poses.indexOf(closestL4Pose)
    }
}